x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/index/IndexResolver.java

```diff
@@ -280,44 +290,79 @@ static IndexResolution mergedMapping(String indexPattern, Map<String, Map<String
+                        }
+                        errorMessage.append("mapped as searchable except in ");
+                        errorMessage.append(Arrays.toString(fieldCap.nonSearchableIndices()));
+                    }
+
                     if (errorMessage.length() > 0) {
-                        errorMessage.append(",");
+                        invalidField = new InvalidMappedField(name, errorMessage.toString());
                     }
-                    errorMessage.append("[" + indexPattern + "] points to indices with incompatible mappings: ");
-                    errorMessage.append("field [" + name + "] is searchable except in ");
-                    errorMessage.append(Arrays.toString(fieldCap.nonSearchableIndices()));
-                }
-                if (errorMessage.length() > 0) {
-                    return IndexResolution.invalid(errorMessage.toString());
                 }
                 
                 // validation passes - create the field
-                // and name wasn't added before
+                // if the name wasn't added before
+                final InvalidMappedField invalidF = invalidField;
+                final FieldCapabilities fieldCapab = fieldCap;
                 if (!flattedMapping.containsKey(name)) {
-                    createField(name, fieldCap, fieldCaps, hierarchicalMapping, flattedMapping, false);
+                    createField(name, fieldCaps, hierarchicalMapping, flattedMapping, s -> {
+                        return invalidF != null ? invalidF : createField(s, fieldCapab.getType(), emptyMap(), fieldCapab.isAggregatable());
+                    });
                 }
             }
         }
 
         return IndexResolution.valid(new EsIndex(indexPattern, hierarchicalMapping));
     }
 
+    private static EsField createField(String fieldName, Map<String, Map<String, FieldCapabilities>> globalCaps,
+            Map<String, EsField> hierarchicalMapping, Map<String, EsField> flattedMapping,
+            Function<String, EsField> field) {
+
+        Map<String, EsField> parentProps = hierarchicalMapping;
+
+        int dot = fieldName.lastIndexOf('.');
+        String fullFieldName = fieldName;
+
+        if (dot >= 0) {
+            String parentName = fieldName.substring(0, dot);
+            fieldName = fieldName.substring(dot + 1);
+            EsField parent = flattedMapping.get(parentName);
+            if (parent == null) {
+                Map<String, FieldCapabilities> map = globalCaps.get(parentName);
+                if (map == null) {
+                    throw new SqlIllegalArgumentException("Cannot find field {}; this is likely a bug", parentName);
+                }
+                FieldCapabilities parentCap = map.values().iterator().next();
+                parent = createField(parentName, globalCaps, hierarchicalMapping, flattedMapping,
+                        s -> createField(s, parentCap.getType(), new TreeMap<>(), parentCap.isAggregatable()));
```

Comment:
maybe I'm overlooking something, but why do you need a new TreeMap here, Collections.emptyMap() won't work?

Reply:
The map is used for the properties of the parent; a treeMap is used since the children need to be sorted out alphabetically. The field caps does not return fields in order anymore and thus this needs to be maintained internally.
`emptyMap` is still used for fields that do not have children but the call above is for a parent (which by definition has).

Reply:
got it, thx!